src/Kestrel.Core/Internal/Http/HttpParser.cs

```diff
@@ -31,445 +32,285 @@ public HttpParser(bool showErrorDetails)
+            if (!reader.IsNext(Eol, advancePast: true))
             {
-                ParseRequestLine(handler, data, span.Length);
+                if (reader.TryRead(out byte value) && value == ByteCR && !reader.TryRead(out value))
+                {
+                    reader.Rewind(reader.Consumed - startPosition);
+
+                    // Incomplete if ends in CR
+                    return false;
+                }
+
+                // Not CR/LF
+                RejectRequestLine(requestLine);
             }
 
-            examined = consumed;
+            ParseRequestLine(handler, requestLine);
             return true;
         }
 
-        private unsafe void ParseRequestLine(TRequestHandler handler, byte* data, int length)
+        private void ParseRequestLine(TRequestHandler handler, in ReadOnlySpan<byte> data)
         {
-            // Get Method and set the offset
-            var method = HttpUtilities.GetKnownMethod(data, length, out var offset);
+            // TODO:
+            // We can simplify GetKnownMethod if we take into account that the
+            // absolute smallest request line is something like "Z * HTTP/1.1"
+            // See https://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1.2
 
-            Span<byte> customMethod = method == HttpMethod.Custom ?
-                GetUnknownMethod(data, length, out offset) :
-                default;
+            ReadOnlySpan<byte> customMethod = default;
 
-            // Skip space
-            offset++;
-
-            byte ch = 0;
-            // Target = Path and Query
-            var pathEncoded = false;
-            var pathStart = -1;
-            for (; offset < length; offset++)
+            // Get Method and set the offset
+            if (!HttpUtilities.GetKnownMethod(data, out HttpMethod method, out int offset))
             {
-                ch = data[offset];
-                if (ch == ByteSpace)
-                {
-                    if (pathStart == -1)
-                    {
-                        // Empty path is illegal
-                        RejectRequestLine(data, length);
-                    }
-
-                    break;
-                }
-                else if (ch == ByteQuestionMark)
-                {
-                    if (pathStart == -1)
-                    {
-                        // Empty path is illegal
-                        RejectRequestLine(data, length);
-                    }
+                customMethod = GetUnknownMethod(data, out offset);
+            }
 
-                    break;
-                }
-                else if (ch == BytePercentage)
-                {
-                    if (pathStart == -1)
-                    {
-                        // Path starting with % is illegal
-                        RejectRequestLine(data, length);
-                    }
+            // Skip the space after the method
+            ReadOnlySpan<byte> target = data.Slice(++offset);
```

Comment:
Split this? Pre-increment can be confusing
```csharp
// Skip the space after the method
offset++;
ReadOnlySpan<byte> target = data.Slice(offset);
```

Reply:
Sure